datajunction-server/alembic/versions/2024_08_18_0036-f3c9b40deb6f_add_create_by_to_nodes_node_revisions_.py

```diff
@@ -0,0 +1,96 @@
+"""add create_by to nodes, node revisions, collections, and tags
+
+Revision ID: f3c9b40deb6f
+Revises: 34171c92dd6d
+Create Date: 2024-08-18 00:36:51.859475+00:00
+
+"""
+# pylint: disable=no-member, invalid-name, missing-function-docstring, unused-import, no-name-in-module
+
+import sqlalchemy as sa
+
+from alembic import op
+
+# revision identifiers, used by Alembic.
+revision = "f3c9b40deb6f"
+down_revision = "34171c92dd6d"
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    """This upgrades adds the created_by_id field to Collection, Node, NodeRevision, and Tag
+
+    Deployments prior to this upgrade will not have easily accessible information for who created
+    these objects in the past. Therefore, this upgrade backfills all existing objects as created by
+    the user with ID=1. If creation information can be gathered elsewhere, it's recommended that you
+    manually backfill the correct created_by_id after the database is upgraded.
+    """
+    with op.batch_alter_table("collection", schema=None) as batch_op:
+        batch_op.add_column(sa.Column("created_by_id", sa.Integer(), nullable=True))
+        batch_op.create_foreign_key(None, "users", ["created_by_id"], ["id"])
+
+    with op.batch_alter_table("node", schema=None) as batch_op:
+        batch_op.add_column(sa.Column("created_by_id", sa.Integer(), nullable=True))
+        batch_op.create_foreign_key(None, "users", ["created_by_id"], ["id"])
+
+    with op.batch_alter_table("noderevision", schema=None) as batch_op:
+        batch_op.add_column(sa.Column("created_by_id", sa.Integer(), nullable=True))
+        batch_op.create_foreign_key(None, "users", ["created_by_id"], ["id"])
+
+    with op.batch_alter_table("tag", schema=None) as batch_op:
+        batch_op.add_column(sa.Column("created_by_id", sa.Integer(), nullable=True))
+        batch_op.create_foreign_key(None, "users", ["created_by_id"], ["id"])
+
+    # Backfill created by columns with user id 1
```

Comment:
Auto increment in postgres starts at 1 (I think) so this would automatically set it to the first user in the system. Even this isn't perfect though since that user could have been deleted, but 0 and -1 would always fail because of the foreign key constraint that requires the ID to exist in the users table. I tried to describe this in the docstring but keep in mind that this is only important for existing deployments that are trying to upgrade forward. New deployments won't have any records to update here and will have the correct user ID set as part of normal operations.

Reply:
Got it. Would it make sense then to create some placeholder user with ID=1 then and call it DJ Root and never allow it to login or be used in APIs?

Reply:
Anyways... up to you.

Reply:
@agorajek that does make sense to me but the thing is it would still require a manual change for existing deployments. This is a one time pain point that we'll fix very quickly for our own deployment. If we make this change to have a dj user with ID 1, it's a lot of code to write and maintain for something that will never be used by future deployments (because it will be impossible).